devito/passes/iet/misc.py

```diff
@@ -74,49 +91,80 @@ def relax_incr_dimensions(iet, **kwargs):
+        roots_max = {i.dim.root: i.symbolic_max for i in outer}
+
+        # A dictionary to map maximum of processed parent dimensions
+        # Helps to neatly handle bounds in hierarchical blocking
+        proc_parents_max = {}
+
+        # Process inner iterations
+        for n, i in enumerate(inner):
+            if i.dim.parent in proc_parents_max and i.symbolic_size == i.dim.parent.step:
+                # In hierarchical blocking (BLOCKLEVELS > 1) a parent dimension may
+                # have already been processed as an 'inner' Iteration. Since in
+                # hierarchical blocking block sizes in lower levels always perfectly
+                # divide block sizes of upper levels we can use the parent's Iteration
+                # maximum
+                iter_max = proc_parents_max[i.dim.parent]
+            else:
+                # Candidate 1: symbolic_max of current iteration
+                # Most of the cases pass though this code:
+                # e.g. `i.symbolic_max = x0_blk0 + x0_blk0_size`
+                symbolic_max = i.symbolic_max
+
+                # Candidate 2: maximum of the root iteration
+                # Candidate 2a: Usually the max of parent/root dimension, e.g. `x_M`
+                root_max = roots_max[i.dim.root]
+
+                # Candidate 2b:
+                # the `symbolic_size` of an Iteration may exceed the size of a parent's
+                # block size (e.g. after CIRE passes) e.g.
+                # `i.dim.parent.step = x0_blk1_size` while
+                # `i.symbolic_size = x0_blk1_size + 4`
+                # For this case, proper margin should be added in order not to omit
+                # iterations
+                lb_margin = i.symbolic_min - i.dim.symbolic_min
+                size_margin = i.symbolic_size - i.dim.parent.step
+                ub_margin = i.dim.parent.symbolic_max + size_margin + lb_margin
+
+                # So candidate 2 is the maximum of the
+                # root's max (2a) and the current iteration's required max (2b)
+                # and instead of `x_M` we may need `x_M + 1` or `x_M + 2`
+
+                # So candidate 2 is
+                # e.g. `domain_max = Max(x_M + 1, x_M)``
+                try:
+                    bool(max(ub_margin, root_max))
+                    domain_max = (max(ub_margin, root_max))
+                except TypeError:
+                    domain_max = MAX(ub_margin, root_max)
+
+                # Finally the iteration's maximum is the minimum of
+                # candidates 1 and 2
+                # e.g. `iter_max = Min(x0_blk0 + x0_blk0_size, domain_max)``
+                try:
+                    bool(min(symbolic_max, domain_max))
+                    iter_max = (min(symbolic_max, domain_max))
+                except TypeError:
+                    iter_max = MIN(symbolic_max, domain_max)
+
+            # Store the selected maximum of this iteration's dimension for
```

Comment:
I suppose this comment is useless, or should be somewhat merged with the one right above the introduction of `proc_parents_max`